---
title: 'Homework for STAT540'
author: 'Sam Hinshaw'
output: 
  html_document:
    toc: TRUE
    toc_depth: 5
    keep_md: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Please view the HTML version of this document, not the GitHub rendering of the `.md` file for proper Table of Contents. 

# Setup
```{r libraries}
# library(BiocInstaller) # or source("https://bioconductor.org/biocLite.R")
suppressPackageStartupMessages({
	library(limma) # biocLite("limma")
	library(edgeR) # biocLite("edgeR")
	library(DESeq2) # biocLite("DESeq2")
	library(dplyr)
	library(magrittr)
	library(devtools)
	library(broom)
	library(ggplot2)
	library(readr)
	library(knitr)
	library(xtable)
	library(pander)
	library(biomaRt)
	library(tidyr)
	library(RColorBrewer)
})
```

Let's check our R version & Packages are up-to-date
We want `R >= 3.2.3`, `limma >= 3.26.7`,  & `edgeR >= 3.12.0`
```{r session info}
sessioninfo <- devtools::session_info()
packages <- sessioninfo$packages
packages %>% 
	filter(package %in% c("limma", "edgeR"))
sessioninfo$platform$version
```

Looks good, let's continue

# Data Inspection

## Download Data

For this, I am using a makefile to avoid downloading the datasets more than once. 
See [Makefile](./Makefile) in this directory. Contents pasted here:
```
all: homework

clean: 
	rm -rf SamHinshawHomework.md SamHinshawHomework.html data_fixed.txt design_fixed.txt

data: 
	curl -o data.txt.gz 'http://stat540-ubc.github.io/homework/assignment/homework_data/NHBE_transcriptome_data.txt.gz?raw=true'
	gunzip -kq data.txt.gz

design: 
	curl -o design.txt 'http://stat540-ubc.github.io/homework/assignment/homework_data/NHBE_design.txt?raw=true'

gitignore: data.txt.gz design.txt data.txt
	grep -q -F 'data.txt' ../.gitignore || echo 'data.txt' >> ../.gitignore
	grep -q -F 'data.txt.gz' ../.gitignore || echo 'data.txt.gz' >> ../.gitignore
	grep -q -F 'design.txt' ../.gitignore || echo 'design.txt' >> ../.gitignore
	grep -q -F 'design_fixed.txt' ../.gitignore || echo 'design_fixed.txt' >> ../.gitignore
	grep -q -F 'data_fixed.txt' ../.gitignore || echo 'data_fixed.txt' >> ../.gitignore

data_fixed: data.txt
	cp data.txt data_fixed.txt
	sed -i -e 's/^"GSE10718_Biomat_1"/"ProbeID"	"GSE10718_Biomat_1"/' data_fixed.txt
	
design_fixed: design.txt
	cp design.txt design_fixed.txt
	sed -i -e 's/^"ExternalID"/"InternalID"	"ExternalID"/' design_fixed.txt
	
homework:  SamHinshawHomework.Rmd gitignore data_fixed design_fixed
	Rscript -e "rmarkdown::render('SamHinshawHomework.Rmd')"
	
```

## Experimental Design

Before we do anything, let's see what's what in our experiment  
<blockquote>
Gene expression patterns were assessed in normal human bronchial epithelial (NHBE) cells exposed to cigarette smoke (CS) from a typical "full flavor" American brand of cigarettes in order to develop a better understanding of the genomic impact of tobacco exposure, which can ultimately define biomarkers that discriminate tobacco-related effects and outcomes in a clinical setting. NHBE cells were treated with CS for 15 minutes and alterations to the transcriptome assessed at 1,2,4 and 24 hours post-CS-exposure using high-density oligonucleotide microarrays.
</blockquote>

## Import Data
```{r read data.txt}
data <- read_delim("data.txt", delim = "\t")
```

Huh, thanks to `readr`, we're getting an error. Something is wrong with our column count. Let's head to the next step and figure out how to handle this error.  

## Inspect Data
```{r check data}
class(data) # First let's make sure this is expected. We should have a data.frame, specifically a tbl_df. 
glimpse(data)
```

It looks like we may have an error with our header, as we've got probe IDs labeled as a sample. Could this be an artifact of the compression? Let's try something else real quick. Also, now we know what dataset we're using! GSE10718.
```{r read data.txt.gz}
data_method2 <- gzfile("data.txt.gz") %>% read_delim(delim = "\t")
```

Nope, looks like it's just a problem with the source file. 
```{r remove data_method2}
rm(data_method2)
```

### Fix Import Errors

Well, since we're not supposed to edit our souce file, I've copied the file do `data_fixed.txt` and appended `"probeIDs	"` to the beginning of the first line with `sed`.  Now we should be ready to rock!
```{r read data_fixed.txt}
data <- read_delim("data_fixed.txt", delim = "\t")
```

Sweet! No errors. Let's take a closer look.
```{r check data_fixed}
glimpse(data)
```

Looking good. These seem to be log2 transformed microarray intensity values.  How about we examine the metadata and then do a quick intensity plot. 
```{r import design}
design <- read_delim("design.txt", delim = "\t")
```

Ugh, same problem again? C'mon guys. Back to `sed`. 
```{r}
design <- read_delim("design_fixed.txt", delim = "\t")
glimpse(design)
```

### Inspect Data

There we go! So we've got 23 samples. How many different treatment groups, and how many different timepoints?
```{r}
design %>% 
	group_by(Treatment) %>% 
	tally()
design %>% 
	group_by(time) %>% 
	tally()
design %>% 
	group_by(Treatment, time) %>% 
	tally()
```

That's a pretty good summary of what's going on, I'd say. Now back to our data file. How many different genes do we have reported?
```{r}
nrow(data)
```

22,737 probes.  Do we know any microarrays that have exactly that many genes?  Fortunately, we don't need to wonder. [This study, GSE10718](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE10718) used [GPL570](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL570), the well known Affymetrix HG U133+ 2.0 Array. Later we can use biomaRt to convert our probeIDs to ensembl IDs (or another equivalent), given the platform information. 

## Basic Data Manipulation
Before we continue, let's create a new column in our dataset that holds our time post-treatment as hours.

```{r}
design %<>% 
	mutate(hours = gsub("\\_h$","", time) %>% as.numeric())
```


## Basic Graphing

However, for now we can just plot some intensity distributions. To help our plot colors out, we can convert our Treatment and time variables to factors
```{r}
design$Treatment <- as.factor(design$Treatment)
design$time <- as.factor(design$time)
```

Now we can `gather()` our data for easy plotting, and even join our design data as well. 
```{r}
gathered_data <- data %>% 
	gather("InternalID", "intensity", 2:24)
glimpse(gathered_data)
gathered_data <- inner_join(gathered_data, design, by = "InternalID")
glimpse(gathered_data)
```

Before we plot, let's set up our color palette.
```{r}
display.brewer.all()
timepoints <- brewer.pal(4, "Set1")
names(timepoints) <- unique(design$time)
treatments <- c("#7FC97F", "#FDC086")
names(treatments) <- unique(design$Treatment)
```

And now, the plot.  First with time-coding...
```{r intensity plot}
p <- ggplot(gathered_data, aes(x = intensity))
p + geom_density(aes(fill = time)) + 
	scale_fill_manual(values = timepoints) + xlab("log2 intensity") + ylab("Frequency") + 
	facet_wrap( ~ InternalID, ncol = 5) + ggtitle("Log2 Intensities of Samples in GSE10718")
```

...and next with treatment coding.
```{r}
p <- ggplot(gathered_data, aes(x = intensity))
p + geom_density(aes(fill = Treatment)) + 
	scale_fill_manual(values = treatments) + xlab("log2 intensity") + ylab("Frequency") + 
	facet_wrap( ~ InternalID, ncol = 5) + ggtitle("Log2 Intensities of Samples in GSE10718")
```

Good for now!  I'll continue later. 

********
This page was last updated on  `r format(Sys.time(), '%A, %B %d, %Y')` at `r format(Sys.time(), '%I:%M%p')`