---
title: 'Homework for STAT540'
author: 'Sam Hinshaw'
output: 
  html_document:
    toc: TRUE
    toc_depth: 5
    keep_md: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Please view the HTML version of this document, not the GitHub rendering of the `.md` file for proper Table of Contents. 

# 0 Setup
```{r libraries}
# library(BiocInstaller) # or source("https://bioconductor.org/biocLite.R")
suppressPackageStartupMessages({
	library(limma) # biocLite("limma")
	library(edgeR) # biocLite("edgeR")
	library(DESeq2) # biocLite("DESeq2")
	library(biomaRt)
	library(yeast2.db) # biocLite("yeast2.db")
	library(hgu133plus2.db) # biocLite("hgu133plus2.db")
	library(magrittr)
	library(devtools)
	# library(broom)
	library(ggplot2)
	library(reshape2)
	library(readr)
	library(knitr)
	library(ggbiplot) # devtools::install_github("vqv/ggbiplot")
	library(FactoMineR) # install.packages("FactoMineR", dependencies = TRUE)
	# library(xtable)
	# library(pander)
	library(tidyr)
	library(RColorBrewer)
	library(viridis)
	library(VennDiagram)
	library(qvalue) #biocLite("qvalue")
	library(dplyr)
	library(cowplot)
	library(GGally)
})
```

Let's check our R version & Packages are up-to-date
We want `R >= 3.2.3`, `limma >= 3.26.7`,  & `edgeR >= 3.12.0`
```{r session info}
sessioninfo <- devtools::session_info()
packages <- sessioninfo$packages
packages %>% 
	dplyr::filter(package %in% c("limma", "edgeR"))
sessioninfo$platform$version
```

Looks good, let's continue

# 1 Data Inspection

## 1.1 Download Data

For this, I am using a makefile to avoid downloading the datasets more than once. 
See [Makefile](./Makefile) in this directory. Contents pasted here:
```
all: homework

clean: 
	rm -rf SamHinshawHomework.md SamHinshawHomework.html data_fixed.txt design_fixed.txt

data: 
	curl -o data.txt.gz 'http://stat540-ubc.github.io/homework/assignment/homework_data/NHBE_transcriptome_data.txt.gz?raw=true'
	gunzip -kq data.txt.gz

design: 
	curl -o design.txt 'http://stat540-ubc.github.io/homework/assignment/homework_data/NHBE_design.txt?raw=true'

gitignore: data.txt.gz design.txt data.txt
	grep -q -F 'data.txt' ../.gitignore || echo 'data.txt' >> ../.gitignore
	grep -q -F 'data.txt.gz' ../.gitignore || echo 'data.txt.gz' >> ../.gitignore
	grep -q -F 'design.txt' ../.gitignore || echo 'design.txt' >> ../.gitignore
	grep -q -F 'design_fixed.txt' ../.gitignore || echo 'design_fixed.txt' >> ../.gitignore
	grep -q -F 'data_fixed.txt' ../.gitignore || echo 'data_fixed.txt' >> ../.gitignore

data_fixed: data.txt
	cp data.txt data_fixed.txt
	sed -i -e 's/^"GSE10718_Biomat_1"/"ProbeID"	"GSE10718_Biomat_1"/' data_fixed.txt
	
design_fixed: design.txt
	cp design.txt design_fixed.txt
	sed -i -e 's/^"ExternalID"/"InternalID"	"ExternalID"/' design_fixed.txt
	
homework:  SamHinshawHomework.Rmd gitignore data_fixed design_fixed
	Rscript -e "rmarkdown::render('SamHinshawHomework.Rmd')"
	
```

### Experimental Design

Before we do anything, let's see what's what in our experiment, [GSE10718](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE10718)
<blockquote>
Gene expression patterns were assessed in normal human bronchial epithelial (NHBE) cells exposed to cigarette smoke (CS) from a typical "full flavor" American brand of cigarettes in order to develop a better understanding of the genomic impact of tobacco exposure, which can ultimately define biomarkers that discriminate tobacco-related effects and outcomes in a clinical setting. NHBE cells were treated with CS for 15 minutes and alterations to the transcriptome assessed at 1,2,4 and 24 hours post-CS-exposure using high-density oligonucleotide microarrays.
</blockquote>

### Import Data
```{r read data.txt}
data <- readr::read_delim("data.txt", delim = "\t")
```

Huh, thanks to `readr`, we're getting an error. Something is wrong with our column count. Let's head to the next step and figure out how to handle this error.  

### Inspect Data
```{r check data}
class(data) # First let's make sure this is expected. We should have a data.frame, specifically a tbl_df. 
glimpse(data)
```

It looks like we may have an error with our header, as we've got probe IDs labeled as a sample. Could this be an artifact of the compression? Let's try something else real quick. Also, now we know what dataset we're using! GSE10718.
```{r read data.txt.gz}
data_method2 <- gzfile("data.txt.gz") %>% readr::read_delim(delim = "\t")
```

Nope, looks like it's just a problem with the source file. 
```{r remove data_method2}
rm(data_method2)
```

### Fix Import Errors

Well, since we're not supposed to edit our souce file, I've copied the file do `data_fixed.txt` and appended `"probeIDs	"` to the beginning of the first line with `sed`.  Now we should be ready to rock!
```{r read data_fixed.txt}
data <- readr::read_delim("data_fixed.txt", delim = "\t")
```

Sweet! No errors. Let's take a closer look.
```{r check data_fixed}
glimpse(data)
```

Looking good. These seem to be log2 transformed microarray intensity values.  How about we examine the metadata and then do a quick intensity plot. 
```{r import design}
design <- read_delim("design.txt", delim = "\t")
```

Ugh, same problem again? C'mon guys. Back to `sed`. 
```{r import design fixed}
design <- readr::read_delim("design_fixed.txt", delim = "\t")
glimpse(design)
```

### Inspect Data

There we go! So we've got 23 samples. How many different treatment groups, and how many different timepoints?
```{r tally}
design %>% 
	group_by(Treatment) %>% 
	tally()
design %>% 
	group_by(time) %>% 
	tally()
design %>% 
	group_by(Treatment, time) %>% 
	tally()
```

That's a pretty good summary of what's going on, I'd say. Now back to our data file. How many different genes do we have reported?
```{r check number of probes}
nrow(data)
```

22,737 probes.  Do we know any microarrays that have exactly that many genes?  Fortunately, we don't need to wonder. This study, [GSE10718](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE10718) used [GPL570](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL570), the well known Affymetrix HG U133+ 2.0 Array. Later we can use biomaRt to convert our probeIDs to ensembl IDs (or another equivalent), given the platform information. 

## 1.2 Basic Data Manipulation
Before we continue, let's create a new column in our dataset that holds our time post-treatment as hours.

```{r add hours column}
design %<>% 
	mutate(hours = gsub("\\_h$","", time) %>% as.numeric())
glimpse(design)
```


## 1.3 Basic Graphing

However, for now we can just plot some intensity distributions. To help our plot colors out, we can convert our Treatment and time variables to factors. This is more for my own benefit (to see intensity distributions) than it is to follow the homework to the letter. Homework progress will continue below. 
```{r create factors}
design$Treatment <- as.factor(design$Treatment)
design$time <- as.factor(design$time)
```

### Intensity Distributions

Let's get biomaRt out of the way so we can move on.  
```{r biomaRt, cache = TRUE, eval = FALSE}
# ensembl <- useMart("ensembl") # set up biomaRt to use the ensembl database
# datasets <- listDatasets(ensembl) # store our datasets so we can see which to use
# ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl) # use the human data set!
# filters <- listFilters(ensembl) # find what filters we can use and cross fingers that it's here
# attributes <- listAttributes(ensembl) # ditto with the attributes
# affy_probe_IDs <- getBM(attributes = c("ensembl_gene_id", "affy_hg_u133_plus_2"), 
# 						filters = "affy_hg_u133_plus_2", 
# 						values = data$ProbeID, 
# 						mart = ensembl, 
# 						uniqueRows = TRUE
# 						) # query the probeIDs
# head(affy_probe_IDs)
# colnames(affy_probe_IDs) <- c("ensembl_gene_id", "ProbeID")
# merged.data <- merge(data, affy_probe_IDs, by = base::intersect(affy_probe_IDs$ProbeID, data$ProbeID), 
# 					 all.y = TRUE, all.x = FALSE, 
# 					 by.x = by, by.y = by)
# 
# joined_probeIDs <- data$ProbeID
# rownames(data) <- data$ProbeID
# glimpse(data)
# 
# affy_probe_IDs_hgnc <- getBM(attributes = c("hgnc_symbol", "affy_hg_u133_plus_2"), 
# 							 filters = "affy_hg_u133_plus_2", 
# 							 values = data$ProbeID, 
# 							 uniqueRows = TRUE,
# 							 mart = ensembl
# 							 ) # query the probeIDs
# colnames(affy_probe_IDs_hgnc) <- c("hgnc_symbol", "ProbeID")
```


```{r}
# affy.gene.sybmols <- mget(data$ProbeID, hgu133plus2SYMBOL)
# # affy.gene.ensembl <- mget(data$ProbeID, hgu133plus2ENSEMBL)
# length(unique(affy.gene.sybmols))
# length(affy.gene.sybmols)
# 
# probe.symbols <- data.frame(
# 	"ProbeID" = names(affy.gene.sybmols),
# 	"geneSymbol" = unlist(affy.gene.sybmols), 
# 	stringsAsFactors = FALSE
# ) %>% tbl_df()
# 
# probe.symbols %<>% 
# 	dplyr::filter(!is.na(geneSymbol))

# 
# probe.geneID <- data.frame(
# 	"ProbeID" = names(affy.gene.ensembl),
# 	"geneID" = unlist(affy.gene.ensembl), 
# 	stringsAsFactors = FALSE
# ) %>% tbl_df()
# 
# probe.geneID %<>% 
# 	dplyr::filter(!is.na(geneID))
# 
# probes.symbols.IDs <- inner_join(probe.symbols, probe.geneID, by = "probeID")

# data <- right_join(data, probe.symbols, by = "ProbeID")
# data.merged <- inner_join(data.merged, probe.geneID, by = "ProbeID")

```

Now we can `gather()` our data for easy plotting, and even join our design data as well. 
```{r gather data}
gathered_data <- data %>% 
	gather("InternalID", "intensity", 2:24)
glimpse(gathered_data)
gathered_data <- inner_join(gathered_data, design, by = "InternalID")
glimpse(gathered_data)
```

Before we plot, let's set up our color palette.
```{r color palette 1, fig.height = 8, fig.width = 4}
display.brewer.all()
timepoints <- brewer.pal(4, "Set1")
names(timepoints) <- unique(design$time)
treatments <- c("#7FC97F", "#FDC086")
names(treatments) <- unique(design$Treatment)
```

And now, the plot.  First with time-coding...
```{r intensity plot faceted time-coded, fig.width = 8, fig.height = 7}
p <- ggplot(gathered_data, aes(x = intensity))
p + geom_density(aes(fill = time)) + 
	scale_fill_manual(values = timepoints) + xlab("log2 intensity") + ylab("Frequency") + 
	facet_wrap( ~ ExternalID, ncol = 5) + ggtitle("Log2 Intensities of Samples in GSE10718")
```

...and then with treatment coding.
```{r intensity plot faceted Treatment-coded, fig.width = 8, fig.height = 7}
p <- ggplot(gathered_data, aes(x = intensity))
p + geom_density(aes(fill = Treatment)) + 
	scale_fill_manual(values = treatments) + xlab("log2 intensity") + ylab("Frequency") + 
	facet_wrap( ~ ExternalID, ncol = 5) + ggtitle("Log2 Intensities of Samples in GSE10718")
```

### Single Probe Intensity

As per the homework, let's plot the gene expression data for a single probe. Let's pick one at pseudo-random!
```{r pick probe}
set.seed(20)
data[sample(nrow(data), 1), 1:6] %>% kable("markdown")
singleprobe <- gathered_data %>% 
	dplyr::filter(ProbeID == "231836_at")
```

```{r plot single probe}
ggplot(data = singleprobe, aes(y = intensity, x = ExternalID)) + 
	geom_bar(stat = "identity") +
	theme(axis.text.x = element_text(angle = 65, hjust = 1))
```

Okay, so we see we've got pretty similar intensities throughout the samples.  But let's group them anyways! First by treatment.
```{r plot single probe treatments}
ggplot(data = singleprobe, aes(y = intensity, x = Treatment)) + 
	geom_bar(stat = "identity", width = 0.5)
```

...then by time
```{r plot single probe times}
ggplot(data = singleprobe, aes(y = intensity, x = time)) + 
	geom_bar(stat = "identity", width = 0.5)
```

Ah-hah! That's rather interesting...
Let's do all permutations of combinations. 
```{r permutations}
gathered_data %<>% mutate(perm = paste(as.character(Treatment), as.character(time), sep = "_"))
singleprobe <- gathered_data %>% 
	dplyr::filter(ProbeID == "231836_at")
```

And now to plot...
```{r plot single probe perms}
ggplot(data = singleprobe, aes(y = intensity, x = perm)) + 
	geom_bar(stat = "identity", width = 0.5) +
	theme(axis.text.x = element_text(angle = 65, hjust = 1))
```

Note that instead of creating new groups to chart, we could have simply color-coded the original bar chart. For example, using varying intensities of colors for time and complementary colors for the treatment. 
```{r color palette 2}
smoke.colors <- brewer.pal(4, "Reds")
control.colors <- brewer.pal(4, "Blues")
names(smoke.colors) <- c("cigarette_smoke_1_h", "cigarette_smoke_2_h", "cigarette_smoke_4_h", "cigarette_smoke_24_h")
names(control.colors) <- c("control_1_h", "control_2_h", "control_4_h", "control_24_h")
perm_palette <- c(control.colors, smoke.colors)
```

```{r reorder factors to make legend pretty}
singleprobe %<>% 
    mutate(IDnum = gsub("^GSM","", ExternalID) %>% as.numeric())
singleprobe$perm <- as.factor(singleprobe$perm)
singleprobe %<>% 
    mutate(perm = reorder(perm, IDnum, max))
```

```{r plot color coded}
ggplot(data = singleprobe, aes(y = intensity, x = ExternalID)) + 
	geom_bar(stat = "identity", aes(fill = perm)) +
	scale_fill_manual(values = perm_palette) + labs(fill = "Conditions") +
	theme(axis.text.x = element_text(angle = 65, hjust = 1))
```


# 2 Assessing Data Quality

Moving on, let's start looking at our data quality, not just visualizing it. 

## 2.1 Heatmaps

```{r}
datacor <- data %>% 
	dplyr::select(-ProbeID) %>% 
	cor()
melted_corr <- melt(datacor) %>% tbl_df()
head(melted_corr)
```

Let's try reordering these factors with `arrange()` to reveal discrepancies. 
```{r}
design$qualitative <- NULL
design %<>% 
	mutate(qualitative = paste0(Treatment, "_", as.character(hours), "_", ExternalID))
colnames(melted_corr) <- c("InternalID", "Var2", "value")
design_for_corr <- design %>% 
	dplyr::select(-ExternalID)
joined_corr <- melted_corr %>% 
	inner_join(design_for_corr, by = "InternalID")
##
design_for_corr2 <- design_for_corr %>% 
	dplyr::select(InternalID, qualitative, hours, Treatment)
colnames(design_for_corr2) <- c("Var2", "qualitative2", "hours2", "Treatment2")
joined_corr %<>% 
	inner_join(design_for_corr2, by = "Var2")
## Rearrange & reorder by time & then treatment
by_time <- joined_corr %>% 
	arrange(-hours, Treatment)
by_time$factororder <- 1:nrow(by_time)
by_time %<>% 
	mutate(qualitative = reorder(qualitative, factororder, max))
by_time %<>% 
	arrange(hours2, Treatment2)
by_time$factororder2 <- 1:nrow(by_time)
by_time %<>% 
	mutate(qualitative2 = reorder(qualitative2, factororder2, max))
## Rearrange & reorder by treatment & then time
by_treatment <- joined_corr %>% 
	arrange(Treatment, -hours)
by_treatment$factororder <- 1:nrow(by_treatment)
by_treatment %<>% 
	mutate(qualitative = reorder(qualitative, factororder, max))
by_treatment %<>% 
	arrange(Treatment2, hours2)
by_treatment$factororder2 <- 1:nrow(by_time)
by_treatment %<>% 
	mutate(qualitative2 = reorder(qualitative2, factororder2, max))
```

```{r}
tilegradient <- brewer.pal(11, "Spectral")
```

Now to plot!
```{r, fig.width = 8, fig.height = 7}
p <- ggplot(joined_corr, aes(x = qualitative2, y = qualitative, fill = value)) + 
	geom_tile() +
	theme(axis.text.x = element_text(angle = 65, hjust = 1)) +
	scale_fill_gradientn(colors = tilegradient)
p
save_plot("heatmap.svg", p, base_height = 7, base_aspect_ratio = 1.14)
ggplot(by_time, aes(x = qualitative2, y = qualitative, fill = value)) + 
	geom_tile() +
	theme(axis.text.x = element_text(angle = 65, hjust = 1)) +
	scale_fill_gradientn(colors = tilegradient)
ggplot(by_treatment, aes(x = qualitative2, y = qualitative, fill = value)) + 
	geom_tile() +
	theme(axis.text.x = element_text(angle = 65, hjust = 1)) +
	scale_fill_gradientn(colors = tilegradient)
```

## 2.2 Outliers

Okay. That's an absolutely horrendous amount of code just to reorder a damn heatmap. I'm never doing that again, I'll stick with my unordered heatmap, which told me what I needed to know. "GSM270874" or "GSE10718-Biomat-10" is an outlier. No significant batch effects noticeable. 

```{r}
outliers <- joined_corr %>% 
	dplyr::filter(InternalID == "GSE10718_Biomat_10")
outliers %<>% 
	mutate(qualitative2 = reorder(qualitative2, value, max))
## Mean Correlation for this sample
outliers %>% 
	dplyr::filter(Var2 != "GSE10718_Biomat_10") %>% 
	summarize(MeanCorr = mean(value))
```

Okay, looks like an outlier, but how can we be sure? What's the average correlation?
```{r}
joined_corr %>% 
	dplyr::filter(value != 1.0) %>% 
	group_by(InternalID) %>% 
	summarize(MeanCorr = mean(value)) %>% 
	arrange(-MeanCorr) %>% 
	kable("markdown")

## What about within groups?

joined_corr %>% 
	dplyr::filter(value != 1.0) %>% 
	group_by(Treatment, hours) %>% 
	summarize(MeanCorr = mean(value)) %>% 
	arrange(-MeanCorr) %>% 
	kable("markdown")
```

Lastly, how does this outlier sample compare to specific treatments (This is a 1hr control sample). For this I'll filter out correlation with itself.
```{r}
outliers %>% 
	dplyr::filter(Var2 != "GSE10718_Biomat_10") %>% 
	group_by(Treatment2, hours2) %>% 
	summarize(MeanCorr = mean(value)) %>% 
	arrange(-MeanCorr) %>% 
	kable("markdown")
```

At last, something meaningful!! This sample correlates MOST those samples within its own group, but also highly with the 1hr cigarette smoke group.  

Quantitatively, it would help to look at mean standard deviation.  

# 3 Differential Expression with Respect to Treatment

## 3.1 Linear Model
```{r}
rownames(data) <- data$ProbeID
fitdata <- data %>% 
	dplyr::select(-ProbeID)
```

```{r}
# design.matrix <- model.matrix(~0+treat.and.hours)
# design.matrix <- model.matrix(~treat.and.hours)
design.matrix <- model.matrix(~0+Treatment, design)
colnames(design.matrix) <- c("CS", "Control")
contrast.matrix <- makeContrasts(CS-Control, levels = design.matrix)
## Let's make sure this looks right:
design.matrix %>% kable("markdown")
fit <- lmFit(fitdata, design.matrix)
fit <- contrasts.fit(fit, contrast.matrix)
efit <- eBayes(fit)
df <- topTable(efit)
df %>% kable("markdown")
results <- decideTests(efit)
vennDiagram(results)
```

In this model, we have ignored time as a covariate, comparing samples against each other just based on treatment group.  In equation terms...
$$Y = X\alpha + \epsilon$$
<center> or </center>
$$Y = \alpha  CigaretteSmoke + \beta  Control + \epsilon$$

Where:  
Y = our responses  
X = our design matrix:   

```{r}
design.matrix %>% kable("markdown")
```

alpha = the parameters of our linear model  
epsilon = the error of our samples  

If that equation isn't formatting correctly, make sure you're viewing the HTML version, not the .md version github has formatted. 


## 3.2 Hits in Linear Model

Okay, so it looks like we've got about 1238 interesting genes. 

First let's take a look at our hits and see how many results we've got with `p < 1e-3`. First we want to see without any adjustment, and then with the FDR adjustment.  
```{r}
tT.treatment <- topTable(efit, adjust = "none", number = Inf, p.value = 1e-3)
# here, we should see that P.Value = adj.P.Val
head(tT.treatment)
nrow(tT.treatment)
tT.treatment$ProbeID <- rownames(tT.treatment)
tT.treatment.sig.fdr <- topTable(efit, adjust = "fdr", number = Inf, 
								 p.value = 0.05, sort.by = "p")
## What is my FDR?
head(tT.treatment.sig.fdr$adj.P.Val, n = 50) %>% sum()
tT.treatment.sig.fdr$ProbeID <- rownames(tT.treatment.sig.fdr)
nrow(tT.treatment.sig.fdr)
head(tT.treatment.sig.fdr)
```

Looks good so far!! Shout out to Louie for explaining that I needed to set up my contrast matrix.  
We've got 805 genes significant at p < 0.001 and 1238 significant genes at p < 0.05 after FDR correction.    Briefly, let's look at our p-value distribution, as recommended by Dr. Pavlidis.  

```{r}
p <- ggplot(tT.treatment.sig.fdr, aes(x = adj.P.Val))
p + geom_histogram(binwidth = 1e-3) + 
	scale_fill_manual(values = timepoints) + xlab("P-Values") + ylab("Frequency") + 
	ggtitle("P-Value Distribution of Significant Genes \n corrected with FDR = 0.05")
```

Looks pretty good!

>*Take the top 50 probes as your “hits” and create a heatmap of their expression levels. Sort the hits by p-values and the samples by treatment.*

Okay, so we've got 1238 hits, let's take the top 50 hits, and then we should go back to our data and filter by the probeIDs. 
```{r}
# gathered_data2 <- data %>% 
# 	gather("InternalID", "intensity", 2:24)
# glimpse(gathered_data2)
# gathered_data2 <- inner_join(gathered_data2, design, by = "InternalID")
# glimpse(gathered_data2)
gathered_data %<>% 
	mutate(qualitative = paste0(Treatment, "_", as.character(hours), "_", ExternalID))
tT.sig.fdr.filtered <- tT.treatment.sig.fdr %>% 
	head(n = 50) %>% tbl_df()
tT.sig.fdr.filtered
topHits <- gathered_data %>% 
	dplyr::filter(ProbeID %in% tT.sig.fdr.filtered$ProbeID)
nrow(topHits)

topHits <- inner_join(topHits, tT.sig.fdr.filtered, by = "ProbeID")
# topHits <- semi_join(topHits, affy_probe_IDs_hgnc, by = "ProbeID")

glimpse(topHits)
topHits$qualitative %<>% as.factor()
topHits %<>% 
	arrange(Treatment, hours)
topHits$treatment.factor <- 1:nrow(topHits)
topHits %<>% 
	mutate(qualitative = reorder(qualitative, treatment.factor, max))
topHits %<>% 
	mutate(hgnc_p.value = paste0(ProbeID, "(", 
								 format(round(adj.P.Val, 4), scientific = TRUE), 
								 ")"))
topHits %<>% 
	arrange(adj.P.Val)
topHits$hgnc_p.value %<>% as.factor()
topHits %<>% 
	mutate(hgnc_p.value = reorder(hgnc_p.value, -adj.P.Val, max))
```


```{r, fig.height = 10, fig.width = 6}
# heatmapcolors <- brewer.pal(9, "RdPu")
# heatmapcolorInterpolate <- colorRampPalette(heatmapcolors)
# heatmapcolorInterpolated <- heatmapcolorInterpolate(nrow(topHits))
ggplot(topHits, aes(x = qualitative, y = hgnc_p.value, fill = intensity)) + 
	geom_tile() +
	theme(axis.text.x = element_text(angle = 65, hjust = 1)) +
	scale_fill_viridis() + xlab("Sample") +
	ylab("Gene Symbol (p-value)")
```

Let's also sort just on time
```{r, fig.height = 10, fig.width = 6}
topHits %<>% 
	arrange(hours, Treatment)
topHits$treatment.factor <- 1:nrow(topHits)
topHits %<>% 
	mutate(qualitative = reorder(qualitative, treatment.factor, max))
ggplot(topHits, aes(x = qualitative, y = hgnc_p.value, fill = intensity)) + 
	geom_tile() +
	theme(axis.text.x = element_text(angle = 65, hjust = 1)) +
	scale_fill_viridis() + xlab("Sample") +
	ylab("Gene Symbol (p-value)")
```

>*What is the (estimated) false discovery rate of this “hits” list? How many of these hits do we expect to be false discoveries?*

Because we've set the p.value of our `topTable` output to 0.05 with an FDR adjustment (`topTable(efit, adjust = "fdr", number = Inf, p.value = 0.05, sort.by = "p")`), we would expect our adjusted values to have an FDR of 0.05.  This means that in a list of 50 genes, we would expect to see 2.5 genes as false discoveries. 

# 4 Differential expression with respect to time

## 4.1 Linear model

Let's pick up where we left off before 3.1, this time setting time in our design matrix.  
```{r}
design.matrix.time <- model.matrix(~hours, design)
# colnames(design.matrix.time) <- c("intercept", "hours")
# contrast.matrix.time <- makeContrasts(hours-intercept, levels = design.matrix.time) # this step is unnecessary!
## Let's make sure this looks right:
design.matrix.time %>% kable("markdown")
fit.time <- lmFit(fitdata, design.matrix.time)
# fit <- contrasts.fit(fit, contrast.matrix.time) # this step is unnecessary!
efit.time <- eBayes(fit.time)
tT.time <- topTable(efit.time)
tT.time %>% kable("markdown")
results.time <- decideTests(efit.time)
vennDiagram(results.time)
```

Interesting so far, let's continue.
```{r}
tT.time <- topTable(efit.time, adjust = "none", number = Inf, p.value = 1e-3)
# here, we should see that P.Value = adj.P.Val
head(tT.time)
nrow(tT.time)

p <- ggplot(tT.time, aes(x = adj.P.Val))
p + geom_histogram(binwidth = 1e-3) + 
	scale_fill_manual(values = timepoints) + xlab("P-Values") + ylab("Frequency") + 
	ggtitle("P-Value Distribution of Significant Genes") + xlim(0, 0.05)
## That's actually not so meaningful since we've filtered on all p-values under 0.001

# topTable.time$ProbeID <- rownames(topTable.time)
tT.time.sig.fdr <- topTable(efit.time, adjust = "fdr", number = Inf, p.value = 0.05)
# topTable.time.sig.fdr$ProbeID <- rownames(topTable.time.sig.fdr)
nrow(tT.time.sig.fdr)
head(tT.time.sig.fdr)

tT.time.sig.fdr %>% arrange(-logFC) %>% head() # as you can see the effect sizes are very small
## so the slopes may be statistically different, but in practice, there's not much of a change. 
## this makes sense because we're not properly modelling the system
## Biologically speaking, we would expect to NOT see a difference in control over time
## but we WOULD expect to see a difference in CS over time
## and by not separating those two groups, we're losing power to detect a slope difference

p <- ggplot(tT.time.sig.fdr, aes(x = adj.P.Val))
p + geom_histogram(binwidth = 1e-3) + 
	scale_fill_manual(values = timepoints) + xlab("P-Values") + ylab("Frequency") + 
	ggtitle("P-Value Distribution of Significant Genes \n corrected with FDR = 0.05")
```

Alright, that's looking much better. We've got 958 hits with p-value under 0.001, and 1451 hits with FDR and p-value under 0.05.  

Let's go straight to question 5 and come back to this one.  

# 5 Differential expression analysis with a full model

## 5.1 Quantify the number of hits for treatment
```{r}
design %<>% 
	arrange(hours)
colnames(data)
# design.matrix.combined <- model.matrix(~0 + hours * Treatment, design) ## this isn't working!
design.matrix.combined <- data.frame(
	row.names = c("GSE10718_Biomat_16", "GSE10718_Biomat_17", "GSE10718_Biomat_19",
					 "GSE10718_Biomat_20", "GSE10718_Biomat_21", "GSE10718_Biomat_22",
					 "GSE10718_Biomat_23", "GSE10718_Biomat_24", "GSE10718_Biomat_4", 
					 "GSE10718_Biomat_5", "GSE10718_Biomat_6", "GSE10718_Biomat_10", 
					 "GSE10718_Biomat_11", "GSE10718_Biomat_12", "GSE10718_Biomat_7", 
					 "GSE10718_Biomat_8", "GSE10718_Biomat_9", "GSE10718_Biomat_13", 
					 "GSE10718_Biomat_14", "GSE10718_Biomat_15", "GSE10718_Biomat_1", 
					 "GSE10718_Biomat_2", "GSE10718_Biomat_3"),
	"int.alpha.cs" = c(1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0), 
	"slope.beta.cs" = c(1, 1, 2, 2, 2, 4, 4, 4, 24, 24, 24, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0),
	"int.alpha.con" = c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1), 
	"slope.beta.con" = c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 2, 2, 2, 4, 4, 4, 24, 24, 24)
	)
## Let's reorder fitdata to be safe we WANT
design %>% arrange(Treatment, hours)

## Looks good
fitdata <- fitdata[c("GSE10718_Biomat_16", "GSE10718_Biomat_17", "GSE10718_Biomat_19",
					 "GSE10718_Biomat_20", "GSE10718_Biomat_21", "GSE10718_Biomat_22",
					 "GSE10718_Biomat_23", "GSE10718_Biomat_24", "GSE10718_Biomat_4", 
					 "GSE10718_Biomat_5", "GSE10718_Biomat_6", "GSE10718_Biomat_10", 
					 "GSE10718_Biomat_11", "GSE10718_Biomat_12", "GSE10718_Biomat_7", 
					 "GSE10718_Biomat_8", "GSE10718_Biomat_9", "GSE10718_Biomat_13", 
					 "GSE10718_Biomat_14", "GSE10718_Biomat_15", "GSE10718_Biomat_1", 
					 "GSE10718_Biomat_2", "GSE10718_Biomat_3")]

## Let's make sure we've got our rownames as well
rownames(fitdata) <- data$ProbeID

## Hopefully this works...
design.matrix.combined %>% kable("markdown")

# colnames(design.matrix.combined) <- c("cs", "control", "hours", "control.hours")
contrast.matrix.combined <- makeContrasts("combined" = slope.beta.cs-slope.beta.con, # combined effect
										  "time" = slope.beta.con, # Time effect alone
										  "treatment" = int.alpha.cs-int.alpha.con, # treatment effect alone
										  levels = design.matrix.combined)

# contrast.matrix.combined <- makeContrasts(control.hours-cs-hours+control, levels = design.matrix.combined)

## Let's make sure this looks right:
fit.combined <- lmFit(fitdata, design.matrix.combined)
fit.combined <- contrasts.fit(fit.combined, contrast.matrix.combined)
efit.combined <- eBayes(fit.combined)
tT.combined <- topTable(efit.combined)
tT.combined %>% head() %>% kable("markdown")
results.combined <- decideTests(efit.combined)
vennDiagram(results.combined)
```

Woohoo! Now I've effectively modeled all 3 biological scenarios.


>*For how many probes is treatment a significant factor at the unadjusted p-value 1e-3, and at FDR 0.05 level?*

```{r}
tT.combined <- topTable(efit.combined, coef = 3, adjust = "none", number = Inf, p.value = 1e-3)
# here, we should see that P.Value = adj.P.Val
head(tT.combined)
nrow(tT.combined)
```

So 621 significant genes at p < 0.001 when looking at combined effect.

```{r}
tT.combined.sig.fdr <- topTable(efit.combined, coef = 3, adjust = "fdr", number = Inf, p.value = 0.05)
head(tT.combined.sig.fdr)
nrow(tT.combined.sig.fdr)
```

And 768 significant genes with FDR correction and p < 0.05.  

Now how about with the combined model?


>*Is this number different from what you reported in 3.2? Why? Quantify the proportion of overlapping probes among your hits, when using the unadjusted p-value threshold of 1e-3.*

Here we have just 573 significantly different genes, down from 805 when just considering treatment (when NOT using FDR correction).  Let's look at the overlap. We'll compare our toptables, `tT.treatment` and `tT.combined`.  
```{r}
tT.treatment %<>% tbl_df()
tT.combined$ProbeID <- rownames(tT.combined)
tT.combined %<>% tbl_df()
intersect(tT.combined$ProbeID, tT.treatment$ProbeID) %>% length()
```

Okay, we've got an overlap of 328 genes, I'd say that's a decent overlap.  

>*Plot the distributions of all the p-values for treatment when using both models, i.e., one from the model in Q3 and one from the full model in this question. Compare and comment on the similarity/differences in the shape of the distributions.*

Sweet, I've already done this! We can run them again though.  I'll plot the p-value distributions when using the FDR correction, as this will more accurately reflect real hits, but I'll make sure to not cap at p < 0.05 this time though.  
```{r}
tT.treatment.fdr <- topTable(efit, adjust = "fdr", number = Inf)

p <- ggplot(tT.treatment.fdr, aes(x = adj.P.Val))
p + geom_histogram(binwidth = 0.01) + 
	scale_fill_manual(values = timepoints) + xlab("P-Values") + ylab("Frequency") + 
	ggtitle("P-Value Distribution for Treatment Model")

tT.combined.fdr <- topTable(efit.combined, coef = 1, adjust = "fdr", number = Inf)

p <- ggplot(tT.combined.fdr, aes(x = adj.P.Val))
p + geom_histogram(binwidth = 0.01) + 
	scale_fill_manual(values = timepoints) + xlab("P-Values") + ylab("Frequency") + 
	ggtitle("P-Value Distribution for Combined Model")
```

The p-value distributions look very similar when compared directly, but it seems as though we may have fewer significant hits when using our combined model.  That doesn't seem right, I would've expected to see a skew towards significance in the combined model, since we're more accurately separating our groups. Let's come back to this later.  

## 5.2 Test the null hypothesis

>*Explain in English what you are modeling with this interaction term (what does it represent?).*
>*For how many probes is the interaction effect significant at the unadjusted p-value 1e-3, and at FDR 0.05 level?*

```{r}
tT.combined <- topTable(efit.combined, coef = 1, adjust = "none", number = Inf, p.value = 1e-3)
# here, we should see that P.Value = adj.P.Val
head(tT.combined)
nrow(tT.combined)
```

So 573 significant genes at p < 0.001 when looking at combined effect.

```{r}
tT.combined.sig.fdr <- topTable(efit.combined, coef = 1, adjust = "fdr", number = Inf, p.value = 0.05)
head(tT.combined.sig.fdr)
nrow(tT.combined.sig.fdr)
```

And 664 significant genes with FDR correction and p < 0.05.  

>*Null hypothesis: there’s no significant interaction between time and treatment.*

With this interaction terms, we are modeling both the slope and intercept of both treatment and control groups.  Then, we can model all three scenarios:  
1. Treatment: Intercept of CS (int.alpha.cs) different than Intercept of Control (int.alpha.con)?  
2. Time: Is the Slope of Control (slope.beta.con) different from zero?  
3. Treatment & Time: Is the Slope of CS (slope.beta.cs) different from Slope of Control (slope.beta.con)?  


## 5.3 Plot a few probes where the interaction does and does not matter 



# Microarray Analysis

## 6.0 Download Data

See Makefile again

## 6.1 Data loading and QC
```{r}
yeast <- readr::read_delim("yeast.tsv", delim = "\t")
glimpse(yeast)
yeast
colnames(yeast)[1] <- "probeID"
gath.yeast <- yeast %>% gather("sample", "intensity", 2:7)
```

We've got 10,928 probes (rows) and 6 samples in what appears to be two groups (b & c).  The first column is our probe ID column.  

First off, let's plot some pairwise comparisons.
```{r, cache = TRUE}
yeast %>% dplyr::select(-probeID) %>% 
	ggpairs(aes(alpha = 0.4)) + theme(
		axis.text.x = element_blank(),
		axis.text.y = element_blank(),
		axis.ticks = element_blank()
	)
```

Beautiful! I'm glad I found the `ggpairs()` function, that's made my life a lot easier.  
We can see easily now, that we've got correlation between two groups:  
- c1, b1, c3  
- c2, b2, b3  

Perhaps we could presmue to group these

| Chemostat? | Batch? |
|------------|--------|
| c1         | **c2** |
| **b1**     | b2     |
| c3         | b3     |

Now let's add in our groups.  I tried to create a phenotype data frame and join, but it kept crashing my computer, so I'm just going to use some regex magic to create our treatment group column.  
```{r}
gath.yeast %<>% 
	mutate(group = gsub("[0-9]$", "", sample))
gath.yeast %<>% 
	mutate(group = gsub("^b$", "batch", group))
gath.yeast %<>% 
	mutate(group = gsub("^c$", "chemostat", group))
gath.yeast
tail(gath.yeast)
```

Looks pretty good!

Let's compute the pearson distance between samples. 
```{r}
yeast.corr <- yeast %>% 
	dplyr::select(-probeID) %>% 
	cor(method = "pearson") %>% 
	as.data.frame()
yeast.corr$correlate <- rownames(yeast.corr)
heatmapcolors <- brewer.pal(9, "Oranges")
yeast.corr %<>% gather("sample", "correlation", 1:6)
ggplot(yeast.corr, aes(x = sample, y = correlate, fill = correlation)) + 
	geom_tile() +
	scale_fill_gradientn(colors = heatmapcolors)
```

Okay, now let's rearrange based on what we decided from our pairwise sample-sample scatterplots and what looks like some obvious clustering in our first heatmap.
```{r}
yeast.corr %<>% 
	mutate(sample = factor(sample, levels = c("b1", "c1", "c3", "c2", "b3", "b2")))
yeast.corr %<>% 
	mutate(correlate = factor(correlate, levels = c("b1", "c1", "c3", "c2", "b3", "b2")))
ggplot(yeast.corr, aes(x = sample, y = correlate, fill = correlation)) + 
	geom_tile() +
	scale_fill_gradientn(colors = heatmapcolors)
```

This makes it seem rather apparent that b1, c1, and c3 are part of one group, and c2, b3, and b2 are part of another group.  Either that or there are some serious flaws with these microarrays.  

Let's plot a heatmap of the the top 100 genes, first unsorted.
```{r, fig.height = 14, fig.width = 6}
gath.yeast %>% 
	group_by(sample) %>% 
	do(head(., n = 100)) %>% 
	ggplot(aes(x = sample, y = probeID, fill = intensity)) + 
	geom_tile() +
	scale_fill_viridis() + xlab("Sample") +
	ylab("probeID")
```

And now with our samples sorted into what we believe to be the true groups.  
```{r, fig.height = 14, fig.width = 6}
gath.yeast$sample %<>% as.factor()
gath.yeast %<>% 
	mutate(sample = factor(sample, levels = c("b1", "c1", "c3", "c2", "b3", "b2")))
gath.yeast %>% 
	group_by(sample) %>% 
	do(head(., n = 100)) %>% 
	ggplot(aes(x = sample, y = probeID, fill = intensity)) + 
	geom_tile() +
	scale_fill_viridis() + xlab("Sample") +
	ylab("probeID")
```

Finally, how about PCA? We can use `FactoMineR::PCA` to get a simple plot, or use `stats::prcomp` for an object we can pipe into `ggplot2`.  
```{r}
noProbes.yeast <- yeast
rownames(noProbes.yeast) <- noProbes.yeast$probeID
noProbes.yeast$probeID <- NULL

fPCA <- FactoMineR::PCA(noProbes.yeast, scale.unit = FALSE)
sPCA <- stats::prcomp(noProbes.yeast, scale. = FALSE)
(PCA.summary <- summary(sPCA)$importance %>% as.data.frame())
```

We can clearly see that b3, b2, and c2 should be grouped together (as previously)

For another quick plot, we can use `geom_point` with `geom_text`.  
```{r}
PCA.rotation <- sPCA$rotation %>% as.data.frame()
PC1.variance <- paste0("PC1, ", round(100*PCA.summary$PC1[2], 2), "% Variance")
PC2.variance <- paste0("PC2, ", round(100*PCA.summary$PC2[2], 2), "% Variance")
ggplot(PCA.rotation, aes(x = PC1, y = PC2, label = rownames(PCA.rotation))) +
	geom_point() + geom_text(nudge_y = 0.05) + xlab(PC1.variance) + ylab(PC2.variance)
```

Note that we get a different looking distribution if we scale our values. Regardless, the clustering is obvious, and our principal components are the same.  
```{r}
sPCA.scaled <- stats::prcomp(noProbes.yeast, scale. = TRUE)
(PCA.scaled.summary <- summary(sPCA.scaled)$importance %>% as.data.frame())
PC1.scaled.variance <- paste0("PC1, ", round(100*PCA.scaled.summary$PC1[2], 2), "% Variance")
PC2.scaled.variance <- paste0("PC2, ", round(100*PCA.scaled.summary$PC2[2], 2), "% Variance")
PCA.rotation.scaled <- sPCA.scaled$rotation %>% as.data.frame()
ggplot(PCA.rotation.scaled, aes(x = PC1, y = PC2, label = rownames(PCA.rotation))) +
	geom_point() + geom_text(nudge_y = 0.05) + xlab(PC1.variance) + ylab(PC2.variance)
```

Finally, if we also wanted to plot our observations, we could use the `ggbiplot` package.  
```{r}
g <- ggbiplot(sPCA, obs.scale = 1, var.scale = 1,
			  alpha = 0.05, circle = TRUE)
g + scale_color_discrete(name = '') +
	theme(legend.direction = 'vertical', 
		  legend.position = 'right') +
	ggtitle("PCA Plot of Yeast Samples")
```

Given all of these plots, we can now satisfactorally conclude that samples b1 and c2 have accidentally been swapped. Let's fix that and continue!

## 6.2 Microarray DEA (2 points)

```{r}
yeast.fixed <- yeast
colnames(yeast.fixed)
colnames(yeast.fixed) <- c("probeID", "c2", "b2", "b3", "c1", "b1", "c3")
yeast.fixed <- yeast.fixed[
	c("probeID",
	  "b1", "b2", "b3", "c1", "c2", "c3")
] 
```

Alrighty, now that we've got our gene IDs set up, let's bring in some qualitative data describing batch vs chemostat.  For this we'll need to gather our data first.  
```{r}
rownames(yeast.fixed) <- yeast.fixed$probeID
yeast.fg <- yeast.fixed %>% # for yeast, fixed & gathered
	gather("sample", "intensity", c(b1, b2, b3, c1, c2, c3))
```

And then we can bring in qualitative data.  We can really just create our own data.frame for this.
```{r}
yeast.fg %<>% 
	mutate(group = gsub("[0-9]$", "", sample))
yeast.fg %<>% 
	mutate(group = gsub("^b$", "batch", group))
yeast.fg %<>% 
	mutate(group = gsub("^c$", "chemostat", group))
yeast.fg
tail(yeast.fg)
```

One last thing before we start with limma, we'll have to create a phenotype data.frame to construct our design matrix. 
```{r}
yeast.design <- data.frame(
	row.names = c("b1", "b2", "b3", "c1", "c2", "c3"),
	"sample" = c("b1", "b2", "b3", "c1", "c2", "c3"), 
	"group"  = c("batch", "batch", "batch", "chemostat", "chemostat", "chemostat")
) %>% tbl_df()
```


Alright, now on to our differential expression analysis (DEA) with limma.  
```{r}
design.matrix.yeast <- model.matrix(~0+group, yeast.design)
colnames(design.matrix.yeast) <- c("Batch", "Chemostat")
contrast.matrix.yeast <- makeContrasts(Batch-Chemostat, levels = design.matrix.yeast)
## Let's make sure this looks right:
design.matrix.yeast %>% kable("markdown")
fit.yeast <- yeast.fixed %>% 
	dplyr::select(-probeID) %>% 
	lmFit(design.matrix.yeast)
fit.yeast <- contrasts.fit(fit.yeast, contrast.matrix.yeast)
efit.yeast <- eBayes(fit.yeast)
tT.yeast <- topTable(efit.yeast, adjust.method="BH", number = Inf) # I know BH is default, but it's still good to specify
#tT.yeast <- topTable(efit.yeast, adjust.method="fdr", number = Inf, p.value = 1e-5) # I know BH is default, but it's still good to specify
## It is not specified to have a cutoff here, so I'm leaving it without a cutoff for now, but still with FDR (aka "BH") correction

tT.yeast %>% head() %>% kable("markdown")
results.yeast <- decideTests(efit.yeast)
vennDiagram(results.yeast)
```

Great! Let's check out our p-value distribution before we continue though
```{r}
p <- ggplot(tT.yeast, aes(x = adj.P.Val))
p + geom_histogram(binwidth = 0.05) + 
	xlab("P-Values") + ylab("Frequency") + 
	ggtitle("P-Value Distribution of Significant \n Genes with BH Correction")
```

That looks pretty drastic, but I'm not sure it's wrong. 
```{r}
tT.yeast$probeID <- rownames(tT.yeast)
(tT.yeast %<>% tbl_df())
```

Here, I'll use biomaRt to get gene IDs and ensembl gene IDs.  We know from the homework that this experiment is using the Affymetrix Yeast Genome Array 2.0 platform, and we can do a quick check on GEO to confirm.  [GSE37599](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE37599), and [GPL2529](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL2529), Affymetrix Yeast Genome 2.0 Array.  Perfect!
```{r biomaRt2, cache = TRUE}
# ensembl <- useMart("ensembl") # set up biomaRt to use the ensembl database
# datasets <- listDatasets(ensembl) # store our datasets so we can see which to use
# ensembl <- useDataset("scerevisiae_gene_ensembl", mart = ensembl) # use the human data set!
# filters <- listFilters(ensembl) # find what filters we can use and cross fingers that it's here
# attributes <- listAttributes(ensembl) # ditto with the attributes
# yeast_probe_IDs <- getBM(attributes = c("external_gene_name", "ensembl_gene_id", "affy_yeast_2"), filters = "affy_yeast_2", values = yeast.fixed$probeID, mart = ensembl) # query the probeIDs
# head(yeast_probe_IDs)
# colnames(yeast_probe_IDs)[3] <- "probeID"
# nrow(yeast_probe_IDs)
# nrow(yeast.fixed)
# yeast.fixed.genes.all <- left_join(yeast.fixed, yeast_probe_IDs, by = "probeID")
# nrow(yeast.fixed.genes.all)
# yeast.fixed.genes <- right_join(yeast.fixed, yeast_probe_IDs, by = "probeID") # inner_join() also works here
# nrow(yeast.fixed.genes) ## Good, 6259 hits is a lot. Let's check to make sure we don't have any NAs in our gene IDs
# sum(is.na(yeast.fixed.genes$external_gene_name))
# ## Good none, but if we look at the data, particularly head(n = 10), we can actually see a missing value, it's just not encoded as "NA". I HATE IMPROPER NA ENCODING
# ## the workaround...
# yeast.fixed.genes %>% 
# 	dplyr::filter(external_gene_name == "") %>% 
# 	nrow()
# ## Ouch, seriously? What about the ensembl gene IDs?
# yeast.fixed.genes %>% 
# 	dplyr::filter(ensembl_gene_id == "") %>% 
# 	nrow()
# ## Okay, that's much better. Before I continue though, I must rectify this NA encoding...
# yeast.fixed.genes$external_gene_name %<>% 
# 	gsub("^$", NA, x = .)
# 
# ## Check our work...
# yeast.fixed.genes %>% 
# 	dplyr::filter(external_gene_name == "") %>% 
# 	nrow()
# sum(is.na(yeast.fixed.genes$external_gene_name))
# ## Much better!
```

Briefly, let's compare biomaRt's ability to match probes with that recommended by the assignment. I believe we should get the same number of hits.
```{r}
# source("http://bioconductor.org/biocLite.R")    # Installing biocLite package
# biocLite()                                      # Load the package 
# biocLite("yeast2.db")                           # Download the package using biocLite
# library("yeast2.db")                            # Load the yeast2.db package 
mget("1772391_at", yeast2ORF)                   # See what gene is that probe for 
gene.ids <- mget(yeast.fixed$probeID, yeast2ORF) # this will generated a named list. Let's make it into a data.frame
gene.ids.df <- data.frame(
	"probeID" = names(gene.ids),
	"geneID" = unlist(gene.ids), 
	stringsAsFactors = FALSE
)
## This is perfect, we'll see where we've got NAs
## Number of Hits:
gene.ids.df %>% 
	dplyr::filter(!is.na(geneID)) %>% 
	nrow()

## Number of whiffs:
gene.ids.df %>% 
	dplyr::filter(is.na(geneID)) %>% 
	nrow()

## Good, well at least we've got more hits than whiffs, but I got more hits with biomaRt.  Let's save our hits. 
gene.id.hits <- gene.ids.df %>% 
	dplyr::filter(!is.na(geneID)) %>% 
	tbl_df()

yeast.genes <- right_join(tT.yeast, gene.id.hits, by = "probeID")
yeast.fg.genes <- right_join(yeast.fg, gene.id.hits, by = "probeID")
yeast.fixed.genes <- right_join(yeast.fixed, gene.id.hits, by = "probeID")
```

Okay, so we've got 5705 hits out of 10928, that's not fantastic, but could be worse.  I got 5768 with biomaRt, but don't feel like dealing with a probe mapping to multiple gene IDs as `uniqueRows = TRUE` just wasn't working, and I couldn't seem to find the right `merge()` arguments (or how to properly use `semi_join()`).

Let's pull out our answers and clean up the table.  I'm deliberately not saving the probes without gene hits, as it was unclear in the homework.
```{r}
yeast.genes %<>% 
	select(-AveExpr, -B)
colnames(yeast.genes) <- c("log.fc", "test.stat", "p.value", "q.value", "probe.id", "gene.id")
write_tsv(yeast.genes, "limma.microarray.results.tsv")
```


## 6.3 Microarray DEA continue

>*Illustrate the differential expression between the batch and the chemostat samples for the top hit (i.e., probe with the lowest p- or q-value).*
```{r}
yeast.genes %<>% 
	arrange(q.value)
head(yeast.genes, n = 1)
```

Alright! Let's look at probe "1772391_at", or YIL057C, a "Protein of unknown function; involved in energy metabolism under respiratory conditions; expression induced under carbon limitation and repressed under high glucose; RGI2 has a paralog, RGI1, that arose from the whole genome duplication."<sup>[1](http://uswest.ensembl.org/Saccharomyces_cerevisiae/Gene/Summary?db=core;g=YIL057C;r=IX:247902-248396;t=YIL057C)</sup>

Let's visualize this probe:
```{r}
yeast.fg %>% 
	filter(probeID == "1772391_at") %>% 
	group_by(group) %>% 
	do(mutate(., mean = mean(intensity))) %>% 
	ggplot(aes(x = group, y = intensity)) + geom_point(stat = "identity", shape = 1, size = 5) + ylim(0, 13) + 
	geom_point(aes(y = mean), shape = 95, size = 12) +
	ggtitle("Intensity of Probe 1772391_at, YIL057C ")
```

>*How many probes are identified as differentially expressed at a false discovery rate (FDR) of 1e-5?*
```{r}
tT.yeast.cutoff <- topTable(efit.yeast, adjust.method="fdr", number = Inf, p.value = 1e-5)
nrow(tT.yeast.cutoff)

```

Here we only get 732 differentially expressed genes. Not too shabby. Let's save this file and move on.
```{r}
tT.yeast.cutoff$probeID <- rownames(tT.yeast.cutoff)
yeast.DE.genes <- inner_join(tT.yeast.cutoff, gene.id.hits, by = "probeID") %>% tbl_df()
yeast.DE.genes %<>% 
	select(-AveExpr, -B)
colnames(yeast.DE.genes) <- c("log.fc", "test.stat", "p.value", "q.value", "probe.id", "gene.id")
write_tsv(yeast.DE.genes, "limma.microarray.results.tsv")
```


# 7 RNA-seq analysis

## 7.1 Load RNA Count Data and Sanity Check

Row names are from Satan!! They are utterly ethereal and can be lost at moment's notice. Fixed with `sed` for the sake of automation.
```{r}
yeast.seq <- readr::read_delim("yeast_counts_fixed.tsv", delim = "\t")
yeast.seq
```

Alright, we've got 6542 genes listed with 6 samples, as before.  The difference between this and our microarray is that here we don't have probes, but actual genes, as identified by stampy aligner and subsequent transcript quantification.  Therefore, you can see we've already got a number of genes similar to our genes as identified by our probe ID to gene ID conversion. A lot of the probes on a microarray are bright or dark controls, and depending on the array (though not the affy array that was used), nonspecific binding probes, and various QC probes used in proprietary "black box" normalization algorithms.  

Anyways, let's continue by checking for the sample ID swap.
```{r}
tilegradient.short <- tilegradient[1:5]
yeast.seq.cor <- yeast.seq %>% 
	dplyr::select(-gene) %>% 
	cor()
yeast.seq.cor %<>% 
	
	reshape2::melt(value.name = "correlation")
ggplot(yeast.seq.cor, aes(x = Var1, y = Var2, fill = correlation)) + 
	geom_tile() +
	scale_fill_gradientn(colors = tilegradient.short)
```

## 7.2 DEA of deep sequencing data

Looks great! No sample ID swap here.  Let's start with edgeR.  Perhaps we should try using the DGElist object. 
```{r}
yeast.seq.dg <- DGEList(counts = yeast.seq[2:7], genes = yeast.seq[1], group = yeast.design$group)
yeast.seq.dg <- calcNormFactors(yeast.seq.dg)
yeast.seq.dg
# yeast.genes <- yeast_probe_IDs %>% dplyr::select(-probeID)
# colnames(yeast.genes) <- c("gene", "ensembl_gene_id")
# yeast.seq.dg$genes <- inner_join(yeast.seq.dg$genes, yeast.genes, by = "gene")
# yeast.seq.dg$genes <- inner_join(yeast.seq.dg$genes, yeast_probe_IDs)
```

Okay, looks good.  

Fortunately we've already got our design matrix!
```{r}
DM.yeast.seq <- model.matrix(~group, yeast.design)
colnames(DM.yeast.seq) <- c("(Intercept)", "chemostat")
rownames(DM.yeast.seq) <- colnames(yeast.seq)[2:7]
DM.yeast.seq
```

Now onto dispersion estimation
```{r}
yeast.seq.dg <- estimateGLMCommonDisp(yeast.seq.dg, DM.yeast.seq)
yeast.seq.dg <- estimateGLMTrendedDisp(yeast.seq.dg, DM.yeast.seq)
yeast.seq.dg <- estimateGLMTagwiseDisp(yeast.seq.dg, DM.yeast.seq)

estimateSizeFactorsForMatrix(yeast.seq.dg$counts)
plotBCV(yeast.seq.dg)
```

```{r, fig.width = 8, fig.height = 4}
yeast.seq.fit <- glmFit(yeast.seq.dg, DM.yeast.seq)
yeast.seq.lrt <- glmLRT(yeast.seq.fit)
tT.yeast.seq <- topTags(yeast.seq.lrt, adjust.method = "fdr", p.value = 1e-5, n = Inf)
nrow(tT.yeast.seq$table)
tT.yeast.seq$table %>% 
	head()
plotSmear(yeast.seq.dg)

yeast.test.deep <- exactTest(yeast.seq.dg)

yeast.test.deep$table$threshold <- as.factor(abs(yeast.test.deep$table$PValue) < 0.05)
ggplot(yeast.test.deep$table, aes(x = logCPM, y = logFC, color = threshold)) +
	geom_point() + scale_color_manual(name = "P Value < 0.05", values = c("#56B4E9", "#FF661D"))

```

Unfortunately, as you can see, the negative binomial method employed by edgeR (and DESeq) does not fit yeast data well, due to the nature of yeast genes.  Therefore, this method is likely not the best for determining hits.  If there is time, I will work on limma-voom in question 9.


Finally, let's pull our topTags results into a specific data.frame as specified by the homework.  
```{r}
edger.deep.results <- tT.yeast.seq$table %>% 
	dplyr::select(-logCPM) %>% 
	tbl_df()
colnames(edger.deep.results) <- c("gene.id", "log.fc", "test.stat", "p.value", "q.value")
write_tsv(edger.deep.results, "edger.deep.results.tsv")
```


## 7.3 DEA of low sequencing data
```{r}
yeast.seq.low <- readr::read_delim("yeast_counts_low_fixed.tsv", delim = "\t")
yeast.seq.low.dg <- DGEList(counts = yeast.seq.low[2:7], genes = yeast.seq.low[1], group = yeast.design$group)
yeast.seq.low.dg <- calcNormFactors(yeast.seq.low.dg)
yeast.seq.low.dg
# yeast.genes <- yeast_probe_IDs %>% dplyr::select(-probeID)
# colnames(yeast.genes) <- c("gene", "ensembl_gene_id")
# yeast.seq.dg$genes <- inner_join(yeast.seq.dg$genes, yeast.genes, by = "gene")
# yeast.seq.dg$genes <- inner_join(yeast.seq.dg$genes, yeast_probe_IDs)
```

Okay, looks good.  

Fortunately we've already got our design matrix!
```{r}
DM.yeast.seq.low <- model.matrix(~group, yeast.design)
colnames(DM.yeast.seq.low) <- c("(Intercept)", "chemostat")
rownames(DM.yeast.seq.low) <- colnames(yeast.seq.low)[2:7]
DM.yeast.seq.low
```

Now onto dispersion estimation
```{r}
yeast.seq.low.dg <- estimateGLMCommonDisp(yeast.seq.low.dg, DM.yeast.seq)
yeast.seq.low.dg <- estimateGLMTrendedDisp(yeast.seq.low.dg, DM.yeast.seq)
yeast.seq.low.dg <- estimateGLMTagwiseDisp(yeast.seq.low.dg, DM.yeast.seq)

estimateSizeFactorsForMatrix(yeast.seq.low.dg$counts)
plotBCV(yeast.seq.low.dg)
```

```{r, fig.width = 8, fig.height = 4}
yeast.seq.low.fit <- glmFit(yeast.seq.low.dg, DM.yeast.seq)
yeast.seq.low.lrt <- glmLRT(yeast.seq.low.fit)
tT.yeast.seq.low <- topTags(yeast.seq.low.lrt, adjust.method = "fdr", p.value = 1e-5, n = Inf)
nrow(tT.yeast.seq.low$table)
tT.yeast.seq.low$table %>% 
	head()

plotSmear(yeast.seq.low.dg)
yeast.test.low <- exactTest(yeast.seq.low.dg)
yeast.test.low$table$threshold <- as.factor(abs(yeast.test.low$table$PValue) < 0.05)
ggplot(yeast.test.low$table, aes(x = logCPM, y = logFC, color = threshold)) +
	geom_point() + scale_color_manual(name = "P Value < 0.05", values = c("#56B4E9", "#FF661D"))

```


And let's write out our results as with deep sequencing!
```{r}
edger.low.results <- tT.yeast.seq.low$table %>% 
	dplyr::select(-logCPM) %>% 
	tbl_df()
colnames(edger.low.results) <- c("gene.id", "log.fc", "test.stat", "p.value", "q.value")
write_tsv(edger.low.results, "edger.low.results.tsv")
```

## 7.4 	Deep vs low sequencing

>*Create a Venn diagram showing all genes identified as differentially expressed (at FDR of 1e-5) in the two previous RNA-Seq analyses and answer the following questions based on the diagram:*

Let's create our Venn Diagram
```{r}
venncolors <- brewer.pal(3, "Set3")[1:2]

yeast.seq.overlap <- calculate.overlap(
	x = list(
		"deep" = edger.deep.results$gene.id, 
		"low"  = edger.low.results$gene.id
	)
)
str(yeast.seq.overlap)
# yeast.seq.overlap %<>% get.venn.partitions()
draw.pairwise.venn(area1 = length(yeast.seq.overlap$a1), 
				   area2 = length(yeast.seq.overlap$a3), 
				   cross.area = length(yeast.seq.overlap$a2), 
				   fill = venncolors, 
				   category = c("Deep", "Low"), 
				   ext.text = FALSE
)
```

>*How many genes were identified by edgeR in both low and deep count data?*

Looks good, seems like our low depth sequencing data is a perfect subset (487) of our deep sequencing (2239).  Therefore, overall we've got just 2240 hits, and it seems like lower depth sequencing just reduces our power to detect differentially expressed genes (albeit perhaps with some other costs).

>*How many genes were identified in all the analyses?*

Okay, let's check the overlap for all of these studies. We can see how many matches we get between the 487 genes from low depth sequencing with the microarray results.  
```{r}
intersect(yeast.DE.genes$gene.id, edger.low.results$gene.id) %>% length()
```

Interesting, about 267 genes match. For fun, let's do another quick Venn Diagram.  

```{r}
venncolors <- brewer.pal(3, "Set3")[1:2]

yeast.seq.overlap <- calculate.overlap(
	x = list(
		"lowseq"  = edger.low.results$gene.id, 
		"microarray" = yeast.DE.genes$gene.id
	)
)
str(yeast.seq.overlap)

draw.pairwise.venn(area1 = length(yeast.seq.overlap$a1), 
				 area2 = length(yeast.seq.overlap$a2), 
				 cross.area = length(yeast.seq.overlap$a3), 
				 fill = venncolors, 
				 category = c("Low Seq", "Microarray"), 
				 ext.text = FALSE
)
```

>*Comment on the effect of sequencing depth on the DEA results.*

As mentioned above, "...it seems like lower depth sequencing just reduces our power to detect differentially expressed genes (albeit perhaps with some other costs)."


# 8 Compare DEA results from RNA-Seq and arrays

## 8.1 Plots of interesting and boring genes

Let's find some genes to look at! We want:  
- 2 genes reported as differentially expressed in BOTH analyses  
- 1 gene reported as differentially expressed in only the microarray assay  
- 1 gene reported as differentially expressed in only the RNA-seq assay  
- 1 gene reported as not differentially expressed in either assay  

To not be overzealous, I'll be pulling results from the low depth sequencing. 
```{r}
set.seed(20)
(our.two.hits <- intersect(yeast.DE.genes$gene.id, edger.low.results$gene.id) %>% 
	base::sample(size = 2)
)
```

Let's fiddle around with our count table to make it ready for plotting.
```{r}
yeast.seq.low.melt <- yeast.seq.low %>% 
	melt() %>% tbl_df()
colnames(yeast.seq.low.melt) <- c("geneID", "sample", "counts")
yeast.seq.low.melt %<>% 
	mutate(group = gsub("[0-9]$", "", sample))
yeast.seq.low.melt %<>% 
	mutate(group = gsub("^b$", "batch", group))
yeast.seq.low.melt %<>% 
	mutate(group = gsub("^c$", "chemostat", group))
yeast.seq.low.melt
tail(yeast.seq.low.melt)
```


```{r}
yeast.fg.genes %>% 
	filter(geneID %in% our.two.hits) %>% 
	group_by(group, geneID) %>% 
	do(mutate(., mean = mean(intensity))) %>% 
	ggplot(aes(x = group, y = intensity)) + geom_point(stat = "identity", shape = 1, size = 5) + ylim(0, 13) + 
	geom_point(aes(y = mean), shape = 95, size = 12) + facet_wrap(~geneID) +
	ggtitle("Intensity of Two Differentially Expressed Genes")
```

```{r}
yeast.seq.low.melt %>% 
	filter(geneID %in% our.two.hits) %>% 
	group_by(group, geneID) %>% 
	do(mutate(., mean = mean(counts))) %>% 
	ggplot(aes(x = group, y = counts)) + geom_point(shape = 1, size = 5) + 
	geom_point(aes(y = mean), shape = 95, size = 12) + facet_wrap(~geneID) +
	ggtitle("Intensity of Two Differentially Expressed Genes")
```


********
This page was last updated on  `r format(Sys.time(), '%A, %B %d, %Y')` at `r format(Sys.time(), '%I:%M%p')`